libcaf_core/caf/detail/parser/fsm.hpp

```diff
@@ -84,8 +84,11 @@
     exit_statement;                                                            \
     goto fsm_after_fin;                                                        \
     s_##name : if (ch == '\0') {                                               \
+      ps.code = pec::success;                                                  \
       exit_statement;                                                          \
-      goto fsm_fin;                                                            \
+      if (ps.code == pec::success)                                             \
+        goto fsm_fin;                                                          \
+      goto fsm_after_fin;                                                      \
```

Comment:
Not super happy about forcing a pessimization to all parsers that don't write to the code. What if, instead of writing to the state, we optionally pick up the return type, i.e., did you consider doing something like this?

```c++
{
  using exit_statement_res = decltype(exit_statement);
  if constexpr (std::is_same_v<exit_statement_res, pec>) {
    ps.code = (exit_statement);
    if (ps.code <= pec::trailing_character)
      goto fsm_fin; 
    goto fsm_after_fin;
  } else {
    exit_statement;
    goto fsm_fin; 
  }
}
```

(extra scope is for `exit_statement_res`)

Reply:
This is a good bridge between the first solution and the second one. Why didn't come up with this? :+1: